text/3389-manifest-lint.md

```diff
@@ -0,0 +1,372 @@
+- Feature Name: `manifest-lint`
+- Start Date: 2023-02-14
+- RFC PR: [rust-lang/rfcs#3389](https://github.com/rust-lang/rfcs/pull/3389)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a `[lints]` table to `Cargo.toml` to configure reporting levels for
+rustc and other tool lints.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, you can configure lints through
+- `#[<level>(<lint>)]` or `#![<level>(<lint>)]`, like `#[forbid(unsafe)]`
+  - But this doesn't scale up with additional targets (benches, examples,
+    tests) or workspaces
+- On the command line, like `cargo clippy -- --forbid unsafe`
+  - This puts the burden on the caller
+- Through `RUSTFLAGS`, like `RUSTFLAGS=--forbid=unsafe cargo clippy`
+  - This puts the burden on the caller
+- In `.cargo/config.toml`'s `target.*.rustflags`
+  - This couples you to the running in specific directories and not running in
+    the right directory causes rebuilds
+  - The cargo team has previously stated that
+    [they would like to see package-specific config moved to manifests](https://internals.rust-lang.org/t/proposal-move-some-cargo-config-settings-to-cargo-toml/13336/14?u=epage)
+
+We would like a solution that makes it easier to share across targets and
+packages for all tools.
+
+See also
+- [rust-lang/rust-clippy#1313](https://github.com/rust-lang/rust-clippy/issues/1313)
+- [rust-lang/cargo#5034](https://github.com/rust-lang/cargo/issues/5034)
+- [EmbarkStudios/rust-ecosystem#59](https://github.com/EmbarkStudios/rust-ecosystem/issues/59)
+- [Proposal: Cargo Lint configuration](https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A new `lints` table would be added to configure lints:
+```toml
+[lints.rust]
+unsafe = "forbid"
+```
+and `cargo` would pass these along as flags to `rustc` and `clippy`.
+
+This would work with
+[RFC 2906 `workspace-deduplicate`](https://rust-lang.github.io/rfcs/2906-cargo-workspace-deduplicate.html?highlight=2906#):
+```toml
+[lints]
+workspace = true
+
+[workspace.lints]
+unsafe = "forbid"
+```
+
+## Documentation Updates
+
+## The `lints` section
+
+*as a new ["Manifest Format" entry](https://doc.rust-lang.org/cargo/reference/manifest.html#the-manifest-format)*
+
+Override the default level of lints from different tools by assigning them to a new level in a
+table, for example:
+```toml
+[lints.rust]
+unsafe = "forbid"
+```
+
+This is short-hand for:
+```toml
+[lints.rust]
+unsafe = { level = "forbid", priority = 0 }
+```
+
+`level` corresponds to the lint levels in `rustc`:
+- `forbid`
+- `deny`
+- `warn`
+- `allow`
+
+`priority` is a signed value that controls which lints or lint groups override other lint groups:
```

Comment:
```suggestion
`priority` is a signed number that controls which lints or lint groups override other lint groups:
```

I also wonder, should it be a float or is restricting it to integers enough?

Reply:
I doubt we need the flexibility of floats.  I've made it explicit about being integers.